cocotb/types/logic.py

```diff
@@ -0,0 +1,205 @@
+from cocotb._py_compat import cache
+
+
+""" DESIGN NOTES
+
+__singleton_cache__
+    Used to ensure there is only ever one instance of a particular value. This is a
+    memory optimization: there will only ever be 4 objects allocated; and also a speed
+    optimization: value equality can be identity equality (as inherited from the `object`
+    base class).
+
+operator `isinstance(other, type(self)`
+    Allows operation between a `Bit` and a `Logic` to return a `Logic`.
+
+strict type equality
+    To use `Logic` or `Bit` in hashable collections we need to define `__hash__` and
+    `__eq__` such that "Hashable objects which compare equal must have the same hash
+    value." (https://hynek.me/articles/hashes-and-equality/, 3/20/21). The best way to
+    achieve this is to make `Logic` and `Bit` never equal, like `tuple` and `list`. If we
+    instead made them equal in hash and value, they would be substitutable; which we
+    don't want.
```

Comment:
Just a comment as to why we don't want substitutability here:
```python
d = {Logic(0)}
assert Bit(0) in d  # passes
```

While technically valid, it would be surprising.